File: leetcode/Python/1464. Maximum Product of Two Elements in an Array.py
import heapq
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        # Solution 1
        '''
        Runtime: 75 ms, faster than 66.82% of Python3 online submissions for Maximum Product of Two Elements in an Array.
        Memory Usage: 14 MB, less than 47.72% of Python3 online submissions for Maximum Product of Two Elements in an Array.

        시간복잡도: O(NlogN) - heap에 삽입하는데 드는 시간복잡도: logN & for문 돌면서 드는 시간: N -> NlogN
        공간복잡도: O(N)
        '''
        heap = []
        for i in nums:
            heapq.heappush(heap, -i)
        
        first = -heapq.heappop(heap)
        second = -heapq.heappop(heap)
        return (first-1)*(second-1)

        # Solution 2
        '''
        Runtime: 68 ms, faster than 75.62% of Python3 online submissions for Maximum Product of Two Elements in an Array.
        Memory Usage: 13.9 MB, less than 47.72% of Python3 online submissions for Maximum Product of Two Elements in an Array.
        
        시간복잡도: O(NlogN) - sort하는데 드는 시간복잡도
        공간복잡도: O(1) -> 새 공간을 쓰지 않고 기존 공간 활용
        '''
        nums.sort(reverse=True)
        first = nums[0]
        second = nums[1]
        return (first-1)*(second-1)

File: leetcode/Java/1464. Maximum Product of Two Elements in an Array.java
import java.util.PriorityQueue;
import java.util.Arrays;
import java.util.Collections;

public class 1464. Maximum Product of Two Elements in an Array {
    class Solution {
        public int maxProduct(int[] nums) {
            
            /* solution 1: sort 이용
            Runtime: 15 ms, faster than 5.27% of Java online submissions for Maximum Product of Two Elements in an Array.
            Memory Usage: 43.8 MB, less than 21.13% of Java online submissions for Maximum Product of Two Elements in an Array.
            
            시간복잡도: O(NlogN) -> Dual-pivot quick sort 사용
            */

            Integer[] list = Arrays.stream(nums).boxed().toArray(Integer[]::new);
            Arrays.sort(list, Collections.reverseOrder());
            int first = list[0];
            int second = list[1];
            return (first-1)*(second-1);

            /* Solution 2: priorityQueue 사용
            Runtime: 5 ms, faster than 24.67% of Java online submissions for Maximum Product of Two Elements in an Array.
            Memory Usage: 44.3 MB, less than 6.44% of Java online submissions for Maximum Product of Two Elements in an Array.
            */
            PriorityQueue<Integer> priorityQueue = new PriorityQueue<>(Collections.reverseOrder());
            for (int i=0; i < nums.length; i++) {
                priorityQueue.add(nums[i]);
            }
            int first = priorityQueue.poll();
            int second = priorityQueue.poll();
            return (first-1)*(second-1);  
        }    
    }
}
